File: XPile 3-17/Programs/test_try.js

function divide(a, b) {
    if ((b === 0)) {
        throw {__ul_type: "ZeroDivisionError", __ul_msg: "division by zero", message: "division by zero"};
    }
    return Math.trunc(a / b);
}

function main() {
    try {
        let result = divide(10, 0);
        console.log("should not reach here");
    } catch (__ex) {
        let __ex_type = __ex && __ex.__ul_type ? __ex.__ul_type : (__ex && __ex.constructor ? __ex.constructor.name : "Error");
        let __ex_msg = __ex && __ex.__ul_msg ? __ex.__ul_msg : (__ex ? String(__ex) : "");
        if (__ex_type === "ZeroDivisionError") {
            console.log("test1: caught division by zero");
        } else {
            throw __ex;
        }
    }
    try {
        throw {__ul_type: "ValueError", __ul_msg: "bad input", message: "bad input"};
    } catch (__ex) {
        let __ex_type = __ex && __ex.__ul_type ? __ex.__ul_type : (__ex && __ex.constructor ? __ex.constructor.name : "Error");
        let __ex_msg = __ex && __ex.__ul_msg ? __ex.__ul_msg : (__ex ? String(__ex) : "");
        if (__ex_type === "ValueError") {
            let e = __ex_msg;
            console.log(("test2: " + e));
        } else {
            throw __ex;
        }
    }
    try {
        console.log("test3: in try");
    } finally {
        console.log("test3: in finally");
    }
    try {
        throw {__ul_type: "RuntimeError", __ul_msg: "something went wrong", message: "something went wrong"};
    } catch (__ex) {
        let __ex_type = __ex && __ex.__ul_type ? __ex.__ul_type : (__ex && __ex.constructor ? __ex.constructor.name : "Error");
        let __ex_msg = __ex && __ex.__ul_msg ? __ex.__ul_msg : (__ex ? String(__ex) : "");
        {
            let e = __ex_msg;
            console.log(("test4: " + e));
        }
    }
    try {
        throw {__ul_type: "TypeError", __ul_msg: "type mismatch", message: "type mismatch"};
    } catch (__ex) {
        let __ex_type = __ex && __ex.__ul_type ? __ex.__ul_type : (__ex && __ex.constructor ? __ex.constructor.name : "Error");
        let __ex_msg = __ex && __ex.__ul_msg ? __ex.__ul_msg : (__ex ? String(__ex) : "");
        if (__ex_type === "ValueError") {
            console.log("test5: not this one");
        } else if (__ex_type === "TypeError") {
            let e = __ex_msg;
            console.log(("test5: " + e));
        } else {
            throw __ex;
        }
    }
    try {
        throw {__ul_type: "TestError", __ul_msg: "oops", message: "oops"};
    } catch (__ex) {
        let __ex_type = __ex && __ex.__ul_type ? __ex.__ul_type : (__ex && __ex.constructor ? __ex.constructor.name : "Error");
        let __ex_msg = __ex && __ex.__ul_msg ? __ex.__ul_msg : (__ex ? String(__ex) : "");
        if (__ex_type === "TestError") {
            console.log("test6: caught");
        } else {
            throw __ex;
        }
    } finally {
        console.log("test6: finally ran");
    }
    try {
        let x = divide(10, 2);
        console.log(("test7: " + String(x)));
    } catch (__ex) {
        let __ex_type = __ex && __ex.__ul_type ? __ex.__ul_type : (__ex && __ex.constructor ? __ex.constructor.name : "Error");
        let __ex_msg = __ex && __ex.__ul_msg ? __ex.__ul_msg : (__ex ? String(__ex) : "");
        if (__ex_type === "ZeroDivisionError") {
            console.log("test7: should not reach");
        } else {
            throw __ex;
        }
    }
    return 0;
}

function __ul_sort(arr, n) { const sub = arr.slice(0, n).sort((a, b) => a < b ? -1 : a > b ? 1 : 0); for (let i = 0; i < n; i++) arr[i] = sub[i]; }

main();

File: XPile 3-17/Programs/test_try.py
from dataclasses import dataclass, field

def __ul_sort(arr, n): arr[:n] = sorted(arr[:n])

class __ul_Exception(Exception):
    def __init__(self, type_name, message):
        self.type_name = type_name
        self.message = message
        super().__init__(message)


def divide(a, b):
    if (b == 0):
        raise __ul_Exception("ZeroDivisionError", 'division by zero')
    return (a // b)

def main():
    try:
        result = divide(10, 0)
        print('should not reach here')
    except Exception as __ex:
        __ex_type = __ex.type_name if isinstance(__ex, __ul_Exception) else type(__ex).__name__
        __ex_msg = __ex.message if isinstance(__ex, __ul_Exception) else str(__ex)
        if __ex_type == "ZeroDivisionError":
            print('test1: caught division by zero')
        else:
            raise
    try:
        raise __ul_Exception("ValueError", 'bad input')
    except Exception as __ex:
        __ex_type = __ex.type_name if isinstance(__ex, __ul_Exception) else type(__ex).__name__
        __ex_msg = __ex.message if isinstance(__ex, __ul_Exception) else str(__ex)
        if __ex_type == "ValueError":
            e = __ex_msg
            print(('test2: ' + e))
        else:
            raise
    try:
        print('test3: in try')
    finally:
        print('test3: in finally')
    try:
        raise __ul_Exception("RuntimeError", 'something went wrong')
    except Exception as __ex:
        __ex_type = __ex.type_name if isinstance(__ex, __ul_Exception) else type(__ex).__name__
        __ex_msg = __ex.message if isinstance(__ex, __ul_Exception) else str(__ex)
        if True:
            e = __ex_msg
            print(('test4: ' + e))
    try:
        raise __ul_Exception("TypeError", 'type mismatch')
    except Exception as __ex:
        __ex_type = __ex.type_name if isinstance(__ex, __ul_Exception) else type(__ex).__name__
        __ex_msg = __ex.message if isinstance(__ex, __ul_Exception) else str(__ex)
        if __ex_type == "ValueError":
            print('test5: not this one')
        elif __ex_type == "TypeError":
            e = __ex_msg
            print(('test5: ' + e))
        else:
            raise
    try:
        raise __ul_Exception("TestError", 'oops')
    except Exception as __ex:
        __ex_type = __ex.type_name if isinstance(__ex, __ul_Exception) else type(__ex).__name__
        __ex_msg = __ex.message if isinstance(__ex, __ul_Exception) else str(__ex)
        if __ex_type == "TestError":
            print('test6: caught')
        else:
            raise
    finally:
        print('test6: finally ran')
    try:
        x = divide(10, 2)
        print(('test7: ' + str(x)))
    except Exception as __ex:
        __ex_type = __ex.type_name if isinstance(__ex, __ul_Exception) else type(__ex).__name__
        __ex_msg = __ex.message if isinstance(__ex, __ul_Exception) else str(__ex)
        if __ex_type == "ZeroDivisionError":
            print('test7: should not reach')
        else:
            raise
    return 0


if __name__ == "__main__":
    main()
